spec.html

```diff
@@ -87,80 +48,59 @@ <h1>Uint8Array.prototype.toHex ( )</h1>
 </emu-clause>
 
 <emu-clause id="sec-uint8array.frombase64">
-  <h1>Uint8Array.fromBase64 ( _value_ [ , _options_ ] )</h1>
+  <h1>Uint8Array.fromBase64 ( _string_ [ , _options_ ] )</h1>
+  <p>The <dfn id="standard-base64-alphabet">standard base64 alphabet</dfn> is a List whose elements are the code points corresponding to every letter and number in the Unicode Basic Latin block along with *"+"* and *"/"*; that is, it is StringToCodePoints(*"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/").</p>
   <emu-alg>
-    1. Let _string_ be ? GetStringForBinaryEncoding(_value_).
+    1. If _string_ is not a String, throw a *TypeError* exception.
     1. Let _opts_ be ? GetOptionsObject(_options_).
     1. Let _alphabet_ be ? Get(_opts_, *"alphabet"*).
     1. If _alphabet_ is *undefined*, set _alphabet_ to *"base64"*.
-    1. Set _alphabet_ to ? ToString(_alphabet_).
+    1. If _alphabet_ is not a String, throw a *TypeError* exception.
     1. If _alphabet_ is neither *"base64"* nor *"base64url"*, throw a *TypeError* exception.
-    1. TODO: normalize whitespace / padding here.
-    1. TODO: figure out what the right defaults for whitespace/padding are.
-    1. Let _characters_ be StringToCodePoints(_string_).
-    1. NOTE: StringToCodePoints is used only because RFC 4648 does not produce a sequence of code units. Implementations may be able to base64-decode _string_ directly.
-    1. If _alphabet_ is *"base64"*, then
-      1. If _characters_ cannot result from applying the base64 encoding specified in section 4 of RFC 4648 to some sequence of bytes, throw a *SyntaxError* exception.
-      1. Let _bytes_ be the unique sequence of bytes such that applying the base64 encoding specified in section 4 of RFC 4648 to that sequence would produce _characters_.
+    1. Let _strict_ be ToBoolean(? Get(_opts_, *"strict"*)).
+    1. NOTE: The order of validation and decoding in the algorithm below is not observable. Implementations are encouraged to perform them in whatever order is most efficient, possibly interleaving validation and stripping whitespace with decoding, as long as the behaviour is observably equivalent.
+    1. Let _input_ be StringToCodePoints(_string_).
+    1. If _alphabet_ is *"base64url"*, then
+      1. If _input_ contains U+002B (PLUS SIGN) or U+002F (SOLIDUS), throw a *SyntaxError* exception.
+      1. Replace all occurrences of U+002D (HYPHEN-MINUS) in _input_ with U+002B (PLUS SIGN).
+      1. Replace all occurrences of U+005F (LOW LINE) in _input_ with U+002F (SOLIDUS).
+    1. NOTE: When _strict_ is *false*, the algorithm below is equivalent to the <a href="https://infra.spec.whatwg.org/#forgiving-base64-decode">forgiving-base64 decode</a> operation in HTML.
+    1. If _strict_ is *false*, then
+      1. Remove all occurrences of U+0009 (TAB), U+000A (LF), U+000C (FF), U+000D (CR), and U+0020 (SPACE) from _input_.
```

Comment:
Observation: The WHATWG algorithm (and thus `atob`) does not consider U+000B VERTICAL TABULATION to be whitespace, but ECMA-262 in general does, as does e.g. [`base64 --decode --ignore-garbage`](https://www.gnu.org/software/coreutils/manual/html_node/base64-invocation.html) (which isn't saying much though, because `--ignore-garbage` discards _any_ non-alphabet character and only U+000A LINE FEED is ignored without it).

Reply:
262 has a much bigger set of whitespace. I'd prefer to match WHATWG unless there's a specific reason to deviate.